### PR TITLE
Exception Handling & Debugging

### PR DESCRIPTION
**Pull request by kariabancroft**
## Description
Main goals:
1. Use exceptions
2. begin/rescue exception blocks
3. create custom exceptions
4. read a stack trace
5. use binding.pry for debugging

My top priority is the exception handling lecture which is scheduled for Wednesday this week, but I also think that the debugging lecture should do this week. I still have one (huge) TODO for that lecture.

## Questions for the Author
- [ ] Have you updated the weekly learning goals in the pedagogy resource - Nope

## Todos
- [x] Interactive exercise for debugging. -- I did add some small methods as an example.
- [x] The SECOND image for the stack trace section is the same as the first. I need to get my pen on it to identify each piece. 

## Feedback Requested
Does this split make sense now?

### CODE REVIEW

File: 02-intermediate-ruby/debugging.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+A stack trace contains a ton of useful information, including a list of methods and blocks that Ruby was inside when the exception was raised. Being able to quickly read a complex stack trace is a super useful skill, so let's examine one now.
+
+First, take a look at the following stack trace that came from running some of the tests within the grocery store project.
+
+![test stack trace](images/test-stack-trace.png)
+
+**Activity**: Now, with your pair, spend a few minutes looking at this error in detail. We know you don't have access to the code, but try anyway to answer the following questions:
+1. What file has the code that was _originally called_ to create the error?
+2. What file has the code where the _error is located_? (Is this different than your answer to the question above?)
```
Comment (tildeee) [inline 02-intermediate-ruby/debugging.md]: These questions (1 & 2) are soOOOoOOO GOOD!!!!! YEAH!!!! THANK YOU!!!

File: 02-intermediate-ruby/debugging.md [hunk 2]
```diff
@@ -0,0 +1,108 @@
+# Debugging
+
+## Learning Goals
+- Read and analyze a stack trace
+- Understand how to use `pry` for debugging in Ruby code
+- Better understand the scope of variables and instances in an application when testing
+
+### Reading Errors via Stack Traces
+
+When an exception is raised, it immediately stops the current method, just like `return`. It will then bubble its way up through the program, method by method. If it makes it all the way to the top, Ruby will print out a summary of the exception and a description of what the program was doing when the exception happened, also known as a **stack trace**.
```
Comment (droberts-ada) [inline 02-intermediate-ruby/debugging.md]: I feel like we are just dropped in here with no context - could you add a short introduction?

File: 02-intermediate-ruby/debugging.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+A stack trace contains a ton of useful information, including a list of methods and blocks that Ruby was inside when the exception was raised. Being able to quickly read a complex stack trace is a super useful skill, so let's examine one now.
+
+First, take a look at the following stack trace that came from running some of the tests within the grocery store project.
+
+![test stack trace](images/test-stack-trace.png)
```
Comment (droberts-ada) [inline 02-intermediate-ruby/debugging.md]: Ruby 2.5 changed the format in which stack traces print. The order is reversed and they have a little more information. Does this image need to be updated?

File: 02-intermediate-ruby/debugging.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+**Review**: While you don't have the code, an error message with a stack trace gives you a lot more information than you might be expecting!
+
+First, let's look at each piece of the provided error.
+
+![test stack trace errors](images/test-stack-trace-markup-p1.png)
+
+Next, we'll take a look at all of the details of the stack trace itself.
+
+![test stack trace details](images/test-stack-trace-markup-p2.png)
```
Comment (droberts-ada) [inline 02-intermediate-ruby/debugging.md]: The dark red in these images is not going to show up well against the black on the projector.

File: 02-intermediate-ruby/debugging.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+Pry also has the following handy commands that let you see what objects are currently available in your program to focus on and interact directly with.
+
+- ``ls``,  shows you the local variables defined in the current context, and any public methods or instance variables defined on the current object.
+- ``cd``, command is used to move into a new object (or scope) inside a Pry session. When inside the new scope it becomes the self for the session and all commands and methods will operate on this new self.
+
+
+### Try it out!
+Open your scrabble or Bank Account project and inspect functionality of your code using PRY! Practice using the ls and cd commands.
+
```
Comment (droberts-ada) [inline 02-intermediate-ruby/debugging.md]: Bank Account -> Grocery Store

File: 02-intermediate-ruby/exception-handling.md [hunk 6]
```diff
@@ -187,7 +145,12 @@ end
 
 Like method or block parameters, the rescued exception is a local variable and may be called anything you please.
 
-### Example: `csv_printer.rb`
+### Exercise
+Setup:
+1. Create a new file called `csv_printer.rb` in your working directory.
+2. Copy the following code into that file.
+3. Review what the code is doing with your seat squad.
```
Comment (droberts-ada) [inline 02-intermediate-ruby/exception-handling.md]: These setup instructions and comprehension Qs are going to be super useful!

File: 02-intermediate-ruby/exception-handling.md [hunk 7]
```diff
@@ -211,44 +174,36 @@ while true
...
-If we expect a method to raise an exception and it doesn't, that's a bug! This means that we must test that our methods raise the exceptions we want, as well as testing nominal behavior. As we saw on in several projects, you can look for an exception in a given block of code in Minitest like this
+**Comprehension Questions:**
+
+Answer the comprehension prior to changing and executing the code to test your understanding.
 
+1. What happens if you remove the `=> exception` code?
+2. What happens if you change the `=> exception` code to `=> my_exception`?
+3. What happens if the `rescue` block is modified to include the code at the end of this loop?
 ```ruby
-it "Raises an ArgumentError when given an invalid word" do
```
Comment (droberts-ada) [inline 02-intermediate-ruby/exception-handling.md]: I do not understand question 3, or what is supposed to be done with the code. Could you clarify?

File: 02-intermediate-ruby/debugging.md (hunk 3 above)
Comment (kariabancroft) [reply 02-intermediate-ruby/debugging.md]: Hmm I guess I haven't upgraded to 2.5 so i'll do that and then see.

File: 02-intermediate-ruby/exception-handling.md (hunk 7 above)
Comment (kariabancroft) [reply 02-intermediate-ruby/exception-handling.md]: Yes - I put the code below, but it wasn't clear the question.

File: 02-intermediate-ruby/debugging.md (hunk 3 above)
Comment (kariabancroft) [reply 02-intermediate-ruby/debugging.md]: Nope - I just ran it using rake on my project and this is how it looks. Is what you're thinking of just the generic stack trace, not an error from minitest?

File: 02-intermediate-ruby/debugging.md (hunk 4 above)
Comment (kariabancroft) [reply 02-intermediate-ruby/debugging.md]: Unfortunately I can just re-edit the images and I'd have to do the edits all over again, so I'm going to leave them for now.

File: 02-intermediate-ruby/debugging.md (hunk 3 above)
Comment (droberts-ada) [reply 02-intermediate-ruby/debugging.md]: Probably. That's kind of annoying that they don't use the new (I would say more readable) format.

**Pull request merged**